### DISCUSSION TITLE
How do I override separateMinorPatch if it is defined globally?

### DISCUSSION THREAD

**Discussion by mrclrchtr** [renovatebot/renovate#8399] [Request Help] [answered]
Hi, unfortunately I have a problem with the "separateMinorPatch" and would be happy about help.
If I have the following configuration, I would expect one PR to be opened for each of the spring minor and patch dependencies, and all other minor and patch dependencies to end up in one PR:

```
  "separateMinorPatch": true,
  "packageRules": [
    {
      "groupName": "non-major",
      "packagePatterns": [
        "*"
      ],
      "updateTypes": [
        "minor",
        "patch"
      ],
      "separateMinorPatch": false
    },
    {
      "groupName": "spring boot",
      "packagePatterns": [
        "^org.springframework.boot:"
      ],
      "updateTypes": [
        "major",
        "minor"
      ]
    },
```

But actually I get one "Update non-major (patch)" and one "Update non-major" for the non-major group. What is wrong with my configuration? 

What I want to achieve is: All minor and patch dependencies should end up in a PR. However, for some specific dependencies I want to open a separate PR for each update type (major, minor, patch). For major updates a separate PR should always be opened.

**Comment by rarkins** [comment]
You can't combine updateTypes and separateMinorPatch in the same package rule. 

Can you describe what it is you are wanting to do?

**Reply by mrclrchtr** [reply]
I would like to achieve the following goals:
- All minor and patch dependencies for which nothing else is configured should be in one "non-major" PR (one catch all minor and patch PR)
- All major dependencies for which nothing else is configured should be in separated PRs (for each major update one PR)
- For certain packages, the minor and patch updates should also be separated into separate PRs. (for each major, minor and patch update one PR)

**Reply by rarkins** [reply]
> * All minor and patch dependencies for which nothing else is configured should be in one "non-major" PR (one catch all minor and patch PR)

```json
    {
      "description": "default non-major updates to a single PR",
      "updateTypes": [
        "minor",
        "patch"
      ],
      "groupName": "non-major",
    },
```

As you want to group minor/patch together, this means the default `separateMinorPatch=false` is adequate.

> * All major dependencies for which nothing else is configured should be in separated PRs (for each major update one PR)

This is default behaviour and not overridden by the above rule, so doesn't need any specific config to achieve

> * For certain packages, the minor and patch updates should also be separated into separate PRs. (for each major, minor and patch update one PR)

```
    {
      "description": "group spring boot packages together by major/minor/patch",
      "packagePatterns": [
        "^org.springframework.boot:"
      ],
      "separateMinorPatch": true,
      "groupName": "spring boot"
    },
```

I think the above rule should be enough to (a) separate minor from patch, and (b) group them by patch/minor/major PRs.

**Reply by mrclrchtr** [reply]
Oh, I have made a mistake in my requirements description.. sorry!
- ~~For certain packages, the minor and patch updates should also be separated into separate PRs. (for each major, minor and patch update one PR)~~
Should have been like this:
- For certain packages, the minor ~~and patch~~ updates should also be separated into separate PRs. (for each major and minor one PR)

So that alle patch updates should always go to the "non-major" PR. 

I just tried to add "patch":
```
{
      "groupName": "spring core",
      "packagePatterns": [
        "^org.springframework:"
      ],
      "separateMinorPatch": true,
      "description": "group spring boot packages together by major/minor/patch",
      "patch": {
        "groupName": "non-major"
      }
    },
```
Then there is a "non-major" PR where everything is collected except for the spring dependencies. For these, a "non-major (patch)" PR is then opened.

Is it possible to move the spring patches to the "non-major" PR.

**Reply by viceice** [reply]
@rarkins you should top post this to mark as answer. 🤗

**Reply by rarkins** [reply]
@mrclrchtr sorry, I'm now quite confused about what exactly you want. My advice:

- Avoid putting config options like `separateMinorPatch` into your top level config if you will also put it into `packageRules`. Instead, use `packageRules` for all
- Recall that all rules are matched, and executed in order. So you can "override" previous rules with newer ones to achieve what you want
- When you name a group, e.g. "foobar", Renovate will default to splitting into major and non-major, but will split into patch/minor/major if you have also specified separateMinorPatch=true

**Reply by mrclrchtr** [reply]
Okay, I'm sorry about that.What I actually want is that all patch and minor updates end up in one PR (non-major). And for individual dependencies I want their minor updates to be separated from the "non-major" PR, but the patches should still end up in it.

I will work with your tips and if I find a solution, I will post it here. Your answer above definitely solves the problem I described at the beginning.

**Comment by viceice** [comment, marked as answer]
Try this:

```json
"packageRules": [
    {
      "description": "default non-major updates to a single PR",
      "updateTypes": [
        "minor",
        "patch"
      ],
      "groupName": "non-major",
    },
   {
      "description": "seperate minor and patch for specific packages",
      "packagePatterns": [
        "^org.springframework.boot:"
      ],
      "separateMinorPatch": true,
    },
    {
      "description": "Use other group for specific minor updates to create separate PR",
      "groupName": "spring boot",
      "packagePatterns": [
        "^org.springframework.boot:"
      ],
      "updateTypes": [
        "minor"
      ]
    },
```

**Reply by mrclrchtr** [reply]
So this is close to the solution, however for spring it still opens an "Update non-major (patch)" PR and the patch updates are not added in the "Update non-major" PR where other minor and patch updates are located. 

This is probably because you can only decide once per packagePattern if it should be split or not or? 

So if I want spring to distinguish between minor and patch, I have to select "separateMinorPatch": true. But if I want the patch updates to be provided with other minor updates in one PR, I have to select "separateMinorPatch": false.

So my requirement is impossible to realize, because splitting into separate PRs is directly related to the "separateMinorPatch" or? And without this flag there is no distinction between patch and minor and patches are always put into one group with minors?

**Reply by mrclrchtr** [reply]
``` 
{
      "description": "default non-major updates to a single PR",
      "groupName": "non-major",
      "updateTypes": [
        "minor",
        "patch"
      ]
    },
    {
      "description":  "Try to use other group for spring minor updates to create separate PR",
      "groupName": "spring boot minor",
      "packagePatterns": [
        "^org.springframework.boot:"
      ],
      "updateTypes": [
        "minor"
      ]
    },
```
This is the result and I would like the spring patch update to be in the update non major PR.
![image](https://user-images.githubusercontent.com/16151098/105628606-29f77a80-5e3e-11eb-80c4-27ed43d4bc3a.png)

**Reply by viceice** [reply]
> You can't combine updateTypes and separateMinorPatch in the same package rule.
> 
> Can you describe what it is you are wanting to do?

@mrclrchtr Yes, your requirement is currently not possible.

**Reply by mrclrchtr** [reply]
Ok that's sad, but thanks anyway for your effort! 

The tool is great and takes a lot of work from us :)
